High availability deployment for VPN (#6978)
gardener/gardener: docs/usage/reversed-vpn-tunnel.md

In the dataplane of the cluster, the `vpn-shoot` will establish the connection to the `vpn-seed-server` indirectly using the SNI API Server endpoint as a http proxy. After the connection has been established requests from the `kube-apiserver` will be handled by the tunnel.

> Please note this feature is still in Beta, so you might see instabilities every now and then.

## High Availability for Reversed VPN Tunnel

Shoots which define `spec.controlPlane.highAvailability.failureTolerance: {node, zone}` get an HA control-plane including a
highly available VPN connection by deploying redundant VPN servers and clients. 

Please note that it is not possible to move an open connection to another VPN tunnel. Especially long-running
commands like `kubectl exec -it ...` or `kubectl logs -f ...` will still break if the routing path must be switched 
because either VPN server or client are not reachable anymore. New request should be possible within seconds.

### HA Architecture for VPN

Establishing a connection from the VPN client on the shoot to the server in the control plane works nearly the same

> way as in the non-HA case. The only difference is that the VPN client targets one of two VPN server, represented by two services 

Reviewer: ```suggestion
way as in the non-HA case. The only difference is that the VPN client targets one of two VPN servers, represented by two services 
```

Author: fixed

`vpn-seed-server-0` and `vpn-seed-server-1` with endpoints in pods with the same name.
The VPN tunnel is used by a `kube-apiserver` to reach nodes, services, or pods in the shoot cluster. 
In the non-HA case, a kube-apiserver uses an HTTP proxy running as a side-car in the VPN server to address

> the shoot networks via the VPN tunnel and the `vpn-shoot` acting as router.

Reviewer: ```suggestion
the shoot networks via the VPN tunnel and the `vpn-shoot` acts as router.
```
?

Author: fixed

In the HA case, the setup is more complicated. Instead of an HTTP proxy in the VPN server, the kube-apiserver has
additional side-cars. One side-car for each VPN client to connect to the corresponding VPN server.

> On the shoot side there are now two `vpn-shoot` pods, each with two VPN clients for each VPN server.

Reviewer 2: ```suggestion
On the shoot side, there are now two `vpn-shoot` pods, each with two VPN clients for each VPN server.
```

Author: fixed

With this setup, there would be four possible routes, but only one can be used. Switching the route kills all
open connections. Therefore, another layer is introduced: link aggregation, also named [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
In Linux, you can create a network link by using several other links as slaves. Bonding is here used with

> active-backup mode. This means the traffic goes only through the active sublink and is only changed if the active

Reviewer 2: ```suggestion
active-backup mode. This means the traffic only goes through the active sublink and is only changed if the active one
```

Author: fixed

becomes unavailable. Switching happens in the bonding network driver without changing any routes. So with this layer, 
vpn-seed-server pods can be rolled without disrupting open connections.

![VPN HA Architecture](images/vpn-ha-architecture.png)

With bonding, there are 2 possible routing paths, ensuring that there is at least one routing path intact even if
one `vpn-seed-server` pod and one `vpn-shoot` pod are unavailable at the same time.

> As it is not possible to use multipath routing, one routing path must be configured explicitly.

Reviewer 2: ```suggestion
As it is not possible to use multi-path routing, one routing path must be configured explicitly.
```

Author: fixed

> Please note this feature is available ONLY for >= 1.18 kubernetes clusters. For clusters with Kubernetes version < 1.18, the default OpenVPN setup will be used by default even if the featuregate is enabled.
> Furthermore, this feature is still in Alpha, so you might see instabilities every now and then. 

## High Availability for Reversed VPN Tunnel

Shoots which define `spec.controlPlane.highAvailability.failureTolerance: {node, zone}` get an HA control-plane including a

> HA for VPN connections by deploying redundant VPN servers and clients. 

Reviewer 2: ```suggestion
highly available VPN connection by deploying redundant VPN servers and clients. 
```

Please note that it is not possible to move an open connection to another VPN tunnel. Especially long-running
commands like `kubectl exec -it ...` or `kubectl logs -f ...` will still break if the routing path must be switched 
because either VPN server or client are not reachable anymore. New request should be possible within seconds.

### HA Architecture for VPN

Establishing a connection from the VPN client on the shoot to the server in the control plane works nearly the same

> way as in the non HA case. The only difference is that the VPN client targets one of two VPN server, represented by two services 

Reviewer 2: ```suggestion
way as in the non-HA case. The only difference is that the VPN client targets one of two VPN server, represented by two services 
```

Author: fixed

`vpn-seed-server-0` and `vpn-seed-server-1` with endpoints in pods with the same name.
The VPN tunnel is used by a `kube-apiserver` to reach nodes, services, or pods in the shoot cluster. 
In the non-HA case, a kube-apiserver uses an HTTP proxy running as a side-car in the VPN server to address
the shoot networks via the VPN tunnel and the `vpn-shoot` acting as router.
In the HA case, the setup is more complicated. Instead of an HTTP proxy in the VPN server, the kube-apiserver has
additional side-cars. One side-car for each VPN client to connect to the corresponding VPN server.
On the shoot side there are now two `vpn-shoot` pods, each with two VPN clients for each VPN server.

> With this setup there would be four possible routes, but only one can be used. Switching the route kills all

Reviewer 2: ```suggestion
With this setup, there would be four possible routes, but only one can be used. Switching the route kills all
```

Author: fixed

> open connections. Therefore another layer is introduced: link aggregation, also named [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).

Reviewer 2: ```suggestion
open connections. Therefore, another layer is introduced: link aggregation, also named [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
```

Author: fixed
